src/lock.rs

```diff
@@ -0,0 +1,174 @@
+pub struct LockedPackage {
+    pub name: PackageName,
+    pub checksum: String,
+    pub repository: String,
+    pub version: Version,
+    pub dependencies: Vec<PackageName>,
+}
+
+impl LockedPackage {
+    /// Captures the source, version and checksum of a Package for use in reproducible installs
+    pub fn lock(package: &Package, repository: String) -> Self {
+        let checksum = digest::digest(&digest::SHA256, &package.tgz);
+
+        Self {
+            name: package.name().to_owned(),
+            repository,
+            checksum: hex::encode(checksum),
+            version: package.version().to_owned(),
+            dependencies: package
+                .manifest
+                .dependencies
+                .iter()
+                .cloned()
+                .map(|d| d.package)
+                .collect(),
+        }
+    }
+
+    /// Validates if another LockedPackage matches this one
+    pub fn validate(&self, other: &Self) -> eyre::Result<()> {
+        ensure!(
+            self.name == other.name,
+            "Package name mismatch - expected {}, actual {}",
+            self.name,
+            other.name
+        );
+        ensure!(
+            self.checksum == other.checksum,
+            "Checksum mismatch - expected {}, actual {}",
+            self.checksum,
+            other.checksum
+        );
+        ensure!(
+            self.repository == other.repository,
+            "Repository mismatch - expected {}, actual {}",
+            self.repository,
+            other.repository
+        );
+        ensure!(
+            self.version == other.version,
+            "Version mismatch - expected {}, actual {}",
+            self.version,
+            other.version
+        );
+        ensure!(
+            self.dependencies == other.dependencies,
+            "Dependencies mismatch"
+        );
+        Ok(())
+    }
+
+    /// Constructs a Dependency instance with matching metadata
+    pub fn as_dependency(&self) -> Dependency {
+        Dependency {
+            package: self.name.clone(),
+            manifest: DependencyManifest {
+                version: VersionReq {
+                    comparators: vec![semver::Comparator {
+                        op: semver::Op::Exact,
+                        major: self.version.major,
+                        minor: Some(self.version.minor),
+                        patch: Some(self.version.patch),
+                        pre: self.version.pre.clone(),
+                    }],
+                },
+                repository: self.repository.clone(),
+            },
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+struct RawLockfile {
+    packages: Vec<LockedPackage>,
+}
+
+/// Captures metadata about currently installed Packages
+///
+/// Used to ensure future installations will deterministically select the exact same packages.
+#[derive(Default)]
+pub struct Lockfile {
+    packages: HashMap<PackageName, Arc<LockedPackage>>,
```

Comment:
why the Arc here? doesn't feel like this is a place which benefits from the sharing - can't you just return an `&LockedPackage` or just clone it?

Reply:
This is a bit residual, originally I was going to implement construction of the dependency graph in parallel and this was a good way to share the locked packages across tasks without cloning. Now it's still helpful in places like https://github.com/helsing-ai/buffrs/blob/mara/lockfile/src/package.rs#L548 but it's not very important. I don't feel strongly against removing it, but I do want to eventually improve concurrency, so it might become relevant again. @mara-schulke what's your take?

Reply:
My sense is that for what you're trying to do here, you care about _concurrency_ but not _parallelism_ - that's to say - you want to kick off your requests to artifactory in parallel but you don't care about running your whole algorithm in parallel.

This is because the dependency resolution algorithm should be defined to be cheaply implementable.

Reply:
Yes, this is what I had in mind, but if you want tasks to have concurrent access to the lockfile (even read-only) you need to either clone it or wrap it in an Arc. That said I think I'll just leave that to the future and keep things simpler for now.